flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/ElasticsearchSinkBase.java

```diff
@@ -211,6 +283,23 @@ public void invoke(T value) throws Exception {
 	}
 
 	@Override
+	public void initializeState(FunctionInitializationContext context) throws Exception {
+		// no initialization needed
+	}
+
+	@Override
+	public void snapshotState(FunctionSnapshotContext context) throws Exception {
+		checkErrorAndRethrow();
+
+		if (flushOnCheckpoint) {
+			do {
+				bulkProcessor.flush();
```

Comment:
Following my arguments above, I think the busy loop you mentioned shouldn't happen, because bulk processor's internal `bulkRequest.numberOfActions()` should always be synced with our `numPendingRecords`. (i.e., it should not occur that `bulkRequest.numberOfActions() == 0` but our own `numPendingRecords != 0`).

So in that case, if `bulkRequest.numberOfActions() == 0` then my original loop implementation just fallbacks to a single pass with 2 condition checks.

To a certain extent, I think it might be better to stick to the original loop implementation, so that we're not locked-in with how the `BulkProcessor`'s flush is implemented. As you can see from a commit I just pushed (2956f99) which modifies the mock bulk processor in tests to correctly mimic the flushing behaviour I described above, the loop implementation still pass the tests.